pkg/ccl/crosscluster/logical/logical_replication_job_test.go

```diff
@@ -1003,38 +1003,207 @@ func TestHeartbeatCancel(t *testing.T) {
 	require.ErrorContains(t, <-retryErrorChan, fmt.Sprintf("replication stream %s is not running, status is STREAM_INACTIVE", prodAID))
 }
 
-func setupLogicalTestServer(
-	t *testing.T, ctx context.Context, clusterArgs base.TestClusterArgs, numNodes int,
+// TestMultipleSourcesIntoSingleDest tests if one destination table can handle
+// conflicts streaming from multiple source tables
+func TestMultipleSourcesIntoSingleDest(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRace(t, "multi cluster/node config exhausts hardware")
+
+	ctx := context.Background()
+
+	clusterArgs := base.TestClusterArgs{
+		ServerArgs: base.TestServerArgs{
+			DefaultTestTenant: base.TestControlsTenantsExplicitly,
+			Knobs: base.TestingKnobs{
+				JobsTestingKnobs: jobs.NewTestingKnobsWithShortIntervals(),
+				DistSQL:          &execinfra.TestingKnobs{},
+			},
+		},
+	}
+
+	numDBs := 3
+	server, s, runners, dbNames := setupServerWithNumDBs(t, ctx, clusterArgs, 1, numDBs)
+	defer server.Stopper().Stop(ctx)
+
+	PGURLs, cleanup := GetPGURLs(t, s, dbNames)
+	defer cleanup()
+
+	jobIDs := make([]jobspb.JobID, 3)
+	for i := range numDBs {
+		// Set a dummy jobID to make the array logic simpler
+		if i == 0 {
+			jobIDs[0] = jobspb.InvalidJobID
+			continue
+		}
+		runners[2].QueryRow(t, "CREATE LOGICAL REPLICATION STREAM FROM TABLE tab ON $1 INTO TABLE tab", PGURLs[i].String()).Scan(&jobIDs[i])
+	}
+
+	// Insert into dest, then check source2 -> dest wins
+	runners[2].Exec(t, "UPSERT INTO tab VALUES (1, 'hello')")
+	runners[1].Exec(t, "UPSERT INTO tab VALUES (1, 'goodbye')")
+	now := s.Clock().Now()
+	WaitUntilReplicatedTime(t, now, runners[2], jobIDs[2])
+	expectedRows := [][]string{
+		{"1", "goodbye"},
+	}
+	runners[2].CheckQueryResults(t, "SELECT * from tab", expectedRows)
+
+	// Write to source1 and source2, which should keep their respective rows but dest should resolve a conflict
+	runners[0].Exec(t, "UPSERT INTO tab VALUES (1, 'insertA')")
+	runners[1].Exec(t, "UPSERT INTO tab VALUES (1, 'insertB')")
+
+	expectedRowsS1 := [][]string{
+		{"1", "insertA"},
+	}
+	expectedRowsDest := [][]string{
+		{"1", "insertB"},
+	}
+	now = s.Clock().Now()
+	WaitUntilReplicatedTime(t, now, runners[2], jobIDs[1])
+	WaitUntilReplicatedTime(t, now, runners[2], jobIDs[2])
+	runners[0].CheckQueryResults(t, "SELECT * from tab", expectedRowsS1)
+	runners[1].CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+	runners[2].CheckQueryResults(t, "SELECT * from tab", expectedRowsDest)
+}
+
+// TestFullyConnectedReplication tests 4 tables that are all streaming
+// from each other and how they handle conflicts
+func TestFullyConnectedReplication(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	skip.UnderRace(t, "multi cluster/node config exhausts hardware")
```

Comment:
nit: this isn't a multi node test.

did you copy this?

Reply:
Didn't copy it, I thought we weren't testing race because it caused WaitUntilReplicatedTime() to timeout. But I see that other tests don't have this so will fix.

Reply:
Fixed